eth_utils/abi.py

```diff
@@ -42,19 +93,564 @@ def collapse_if_tuple(abi: Dict[str, Any]) -> str:
+
+    return None
+
+
+def _log_entry_data_to_bytes(
+    log_entry_data: Union[Primitives, HexStr, str],
+) -> HexBytes:
+    return hexstr_if_str(to_bytes, log_entry_data)
+
+
+def _raise_mismatched_abi_error(
+    function_identifier: str,
+    matching_function_signatures: Sequence[str],
+    arg_count_matches: int,
+    encoding_matches: int,
+    args: Optional[Sequence[Any]] = None,
+    kwargs: Optional[Any] = None,
+) -> None:
+    if arg_count_matches == 0:
+        diagnosis = "\nFunction invocation failed due to improper number of arguments."
+    elif encoding_matches == 0:
+        diagnosis = "\nFunction invocation failed due to no matching argument types."
+    elif encoding_matches > 1:
+        diagnosis = (
+            "\nAmbiguous argument encoding. "
+            "Provided arguments can be encoded to multiple functions "
+            "matching this call."
+        )
+
+    if args is not None:
+        collapsed_args = _abi_inputs_types(args)
+    else:
+        collapsed_args = ""
+
+    if kwargs is not None:
+        collapsed_kwargs = dict(
+            {(k, _abi_inputs_types([v])) for k, v in kwargs.items()}
+        )
+    else:
+        collapsed_kwargs = {}
+
+    raise MismatchedABI(
+        f"\nCould not identify the intended function with name "
+        f"`{function_identifier}`, positional arguments with type(s) "
+        f"`{collapsed_args}` and keyword arguments with type(s) "
+        f"`{collapsed_kwargs}`."
+        f"\nFound {len(matching_function_signatures)} function(s) with the name "
+        f"`{function_identifier}`: {matching_function_signatures}{diagnosis}"
+    )
+
+
+def get_all_function_abis(abi: ABI) -> ABIFunction:
+    """
+    Return interfaces for each function in the contract ABI.
+
+    :param abi: Contract ABI.
+    :param type: `ABI`
+    :return: List of ABIs for each function interface.
+    :rtype: `list[ABIFunction]`
+    """
+    return cast(ABIFunction, _filter_by_type("function", abi))
+
+
+def get_function_abi(
+    abi: ABI,
+    function_identifier: Optional[str] = None,
+    args: Optional[Sequence[Any]] = None,
+    kwargs: Optional[Any] = None,
+    abi_codec: Optional[Any] = None,
+) -> ABIFunction:
+    """
+    Return the interface for a contract function.
+
+    :param abi: Contract ABI.
+    :param type: `ABI`
+    :param function_identifier: Find a function ABI with matching name.
+    :param type: `str`
+    :param args: Find a function ABI with matching args.
+    :param type: `list[Any]`
+    :param kwargs: Find a function ABI with matching kwargs.
+    :param type: `Any`
+    :param abi_codec: Codec used for encoding and decoding. Default with
+    `strict_bytes_type_checking` enabled.
+    :param type: `Any`
+    :return: ABI for the function interface.
+    :rtype: `ABIFunction`
+    """
+    if function_identifier is None or not is_text(function_identifier):
+        raise TypeError("Unsupported function identifier")
+
+    if abi_codec is None:
```

Comment:
I'm not sure where and how to document this feature. Maybe the docstring should point to the codec docs?

Reply:
That makes sense to me.